pinecone/index_connection_test.go

```diff
@@ -243,6 +244,196 @@ func (ts *IndexConnectionTests) TestListVectors() {
+	tests := []struct {
+		name  string
+		input FetchVectorsResponse
+		want  string
+	}{
+		{
+			name: "All fields present",
+			input: FetchVectorsResponse{
+				Vectors: map[string]*Vector{
+					"vec-1": {Id: "vec-1", Values: []float32{0.01, 0.01, 0.01}},
+					"vec-2": {Id: "vec-2", Values: []float32{0.02, 0.02, 0.02}},
+				},
+				Usage: &Usage{ReadUnits: toUInt32(5)},
+			},
+			want: `{"vectors":{"vec-1":{"id":"vec-1","values":[0.01,0.01,0.01]},"vec-2":{"id":"vec-2","values":[0.02,0.02,0.02]}},"usage":{"read_units":5}}`,
+		},
+		{
+			name:  "Fields omitted",
+			input: FetchVectorsResponse{},
+			want:  `{}`,
+		},
+		{
+			name: "Fields empty",
+			input: FetchVectorsResponse{
+				Vectors: nil,
+				Usage:   nil,
+			},
+			want: `{}`,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			bytes, err := json.Marshal(tt.input)
+			if err != nil {
+				t.Fatalf("Failed to marshal FetchVectorsResponse: %v", err)
+			}
+
+			if got := string(bytes); got != tt.want {
+				t.Errorf("Marshal FetchVectorsResponse got = %s, want = %s", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestMarshalingListVectorsResponse(t *testing.T) {
+	vectorId1 := "vec-1"
+	vectorId2 := "vec-2"
+	paginationToken := "next-token"
+	tests := []struct {
+		name  string
+		input ListVectorsResponse
+		want  string
+	}{
+		{
+			name: "All fields present",
+			input: ListVectorsResponse{
+				VectorIds:           []*string{&vectorId1, &vectorId2},
+				Usage:               &Usage{ReadUnits: toUInt32(5)},
+				NextPaginationToken: &paginationToken,
+			},
+			want: `{"vector_ids":["vec-1","vec-2"],"usage":{"read_units":5},"next_pagination_token":"next-token"}`,
+		},
+		{
+			name:  "Fields omitted",
+			input: ListVectorsResponse{},
+			want:  `{}`,
+		},
+		{
+			name: "Fields empty",
+			input: ListVectorsResponse{
+				VectorIds:           nil,
+				Usage:               nil,
+				NextPaginationToken: nil,
+			},
+			want: `{}`,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			bytes, err := json.Marshal(tt.input)
+			if err != nil {
+				t.Fatalf("Failed to marshal ListVectorsResponse: %v", err)
+			}
+
+			if got := string(bytes); got != tt.want {
+				t.Errorf("Marshal ListVectorsResponse got = %s, want = %s", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestMarshalingQueryVectorsResponse(t *testing.T) {
+	tests := []struct {
+		name  string
+		input QueryVectorsResponse
+		want  string
+	}{
+		{
+			name: "All fields present",
+			input: QueryVectorsResponse{
+				Matches: []*ScoredVector{
+					{Vector: &Vector{Id: "vec-1", Values: []float32{0.01, 0.01, 0.01}}, Score: 0.1},
+					{Vector: &Vector{Id: "vec-2", Values: []float32{0.02, 0.02, 0.02}}, Score: 0.2},
+				},
+				Usage: &Usage{ReadUnits: toUInt32(5)},
+			},
+			want: `{"matches":[{"vector":{"id":"vec-1","values":[0.01,0.01,0.01]},"score":0.1},{"vector":{"id":"vec-2","values":[0.02,0.02,0.02]},"score":0.2}],"usage":{"read_units":5}}`,
```

Comment:
Might be worth at least having the `want` values as globals at the top? Or fixtures or whatever? Just to downsize the amount of duplicated code

Reply:
It's definitely a lot of duplicative code. I'd prefer not to move the `want` blocks around though as they're specific to the struct that's being tested, and it'd probably be more confusing to have to look them up elsewhere than in-line.

I can try and think of ways to clean some of this up though.